code/modules/mob/living/carbon/human/species_types/mothmen.dm

```diff
@@ -47,3 +55,125 @@
 	if(istype(weapon, /obj/item/melee/flyswatter))
 		return 9 //flyswatters deal 10x damage to moths
 	return 0
+
+/datum/species/moth/on_species_gain(mob/living/carbon/human/H)
+	..()
+	cocoon_action = new()
+	cocoon_action.Grant(H)
+
+/datum/species/moth/on_species_loss(mob/living/carbon/human/H)
+	..()
+	cocoon_action.Remove(H)
+	QDEL_NULL(cocoon_action)
+/*
+/datum/species/moth/spec_WakeUp(mob/living/carbon/human/H)
+	if(H.has_status_effect(STATUS_EFFECT_COCOONED))
+		return TRUE //Cocooned mobs dont get to wake up
+*/
+/datum/action/innate/cocoon
+	name = "Cocoon"
+	desc = "Restore your wings and antennae, and heal some damage. If your cocoon is broken externally you will take heavy damage!"
+	check_flags = AB_CHECK_RESTRAINED|AB_CHECK_STUN|AB_CHECK_CONSCIOUS
+	button_icon_state = "wrap_0"
+	icon_icon = 'icons/mob/actions/actions_animal.dmi'
+
+//
+/datum/action/innate/cocoon/Activate()
+	var/mob/living/carbon/H = owner
+	var/obj/item/organ/wingcheck = H.getorgan(/obj/item/organ/wings/moth)
+	if(!wingcheck) //This is to stop easy organ farms
+		to_chat(H, "<span class='warning'>You don't have any wings to regenerate!</span>")
+		return
+	if(!HAS_TRAIT(H, TRAIT_MOTH_BURNT))
+		to_chat(H, "<span class='warning'>Your wings are fine as they are!</span>")
+		return
+	if(H.nutrition < COCOON_NUTRITION_AMOUNT)
+		to_chat(H, "<span class='warning'>You are too hungry to weave a cocoon!</span>")
+		return
+	H.visible_message("<span class='notice'>[H] begins to hold still and concentrate on weaving a cocoon...</span>", \
+	"<span class='notice'>You begin to focus on weaving a cocoon... (This will take [COCOON_WEAVE_DELAY / 10] seconds and you must hold still.)</span>")
+	H.adjustStaminaLoss(20, 0) //this is here to deter people from spamming it if they get interrupted
+	if(do_after(H, COCOON_WEAVE_DELAY, FALSE, H))
+		if(!ismoth(H))
+			to_chat(H, "<span class='warning'>You have lost your mandibles and cannot weave anymore!.</span>")
```

Comment:
incredibly cursed lmao, how would this happen, but fine

also check if they have the trait still

Reply:
> how would this happen

Mutation toxin shenanigans
fair point about checking for the trait